### ISSUE TITLE
Can't concatenate `ParamSpec` and `TypeVarTuple`

### ISSUE THREAD

**Issue by thomkeh** [microsoft/pyright#4342]
**Is your feature request related to a problem? Please describe.**
I'm trying to implement a type-safe version of `partial`.

This is what I have
```python
from typing import Callable, Concatenate, Generic, ParamSpec, TypeVar, TypeVarTuple, reveal_type

P = ParamSpec("P")
R_co = TypeVar("R_co", covariant=True)
Ts = TypeVarTuple("Ts")

class simple_partial(Generic[P, R_co, *Ts]):
    def __init__(self, __func: Callable[Concatenate[*Ts, P], R_co], *args: *Ts):
        self.func = __func
        self.args = args

    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R_co:
        return self.func(*self.args, *args, **kwargs)

def f(x: int, y: str, z: bool | None = None) -> float:
    print(x, y, z)
    return 0.0

g = simple_partial(f, 2)
reveal_type(g.args)
reveal_type(g.func)
reveal_type(g.__call__)

g(y="foo")  # OK
g(3, z=True)  # should be an error
```

pyright's output:
```
Found 1 source file
pyright 1.1.283
better_partial.py
  better_partial.py:13:27 - error: Expected 1 positional argument (reportGeneralTypeIssues)
  better_partial.py:19:20 - error: Argument of type "(x: int, y: str, z: bool | None = None) -> float" cannot be assigned to parameter "__func" of type "(*Ts@simple_partial, **P@simple_partial) -> R_co@simple_partial" in function "__init__"
    Type "(x: int, y: str, z: bool | None = None) -> float" cannot be assigned to type "(*tuple[int], y: str, z: bool | None = None) -> float"
      Parameter 1: type "*tuple[int]" cannot be assigned to type "int"
        "*tuple[int]" is incompatible with "int" (reportGeneralTypeIssues)
  better_partial.py:20:13 - information: Type of "g.args" is "tuple[Unknown]"
  better_partial.py:21:13 - information: Type of "g.func" is "(Unknown, ...) -> Unknown"
  better_partial.py:22:13 - information: Type of "g.__call__" is "(...) -> Unknown"
2 errors, 0 warnings, 3 informations 
```

I think this should in theory be correct (though maybe I made a mistake). It does work if I used a plain `TypeVar` instead of a `TypeVarTuple`.

**Comment by erictraut** [comment]
`TypeVarTuple` and `ParamSpec` were not designed to work together this way. The behavior would be ambiguous because there's no way to know how many parameters to capture with the TypeVarTuple versus the ParamSpec. 

Using your example, function `f` has three parameters (`x`, `y`, and `z`). Which of these did you intend to be captured by the `*Ts` and which of these did you expect to be captured by `P`? Both `*Ts` and `P` are variadic, so there's no way to resolve this ambiguity.

**Comment by erictraut** [comment]
Incidentally, pyright already contains special-case support for `functools.partial`. So does mypy. It can't be described using the standard type system, so it requires special-case logic inside the type checker.

**Comment by thomkeh** [comment]
>Which of these did you intend to be captured by the *Ts and which of these did you expect to be captured by P?

Isn't this determined by whatever is passed to `*args: *Ts`?

**Comment by thomkeh** [comment]
pyright has no problem with this, but it's the same situation:

```python
Ts = TypeVarTuple("Ts")
Us = TypeVarTuple("Us")
def h(x: tuple[*Ts, *Us], *args: *Ts) -> tuple[*Us]: ...

x = h((0, "foo"), 0)
reveal_type(x)  # Type of "x" is "tuple[Literal['foo']]"
```

**Comment by erictraut** [comment]
Pyright emits an error in this code because PEP 646 explicitly disallows a tuple from containing more than one unpacked TypeVarTuple. This was disallowed for the same reason — that it creates ambiguities when there are multiple variadics to match against.

I applaud your creativity in trying to concatenate a TypeVarTuple and a ParamSpec, but this isn't going to work. You'll need to look for a different solution.
